components/identity.oauth.uma.permission.service/src/main/java/org/wso2/carbon/identity/oauth/uma/service/dao/PermissionTicketDAO.java

```diff
@@ -0,0 +1,165 @@
+            UMAResourceException, PermissionDAOException {
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection()) {
+            checkResourceIdsExistence(connection, resourceList);
+            checkResourceScopesExistence(connection, resourceList);
+            connection.setAutoCommit(false);
+            try (PreparedStatement preparedStatement = connection.prepareStatement(STORE_PT_QUERY)) {
+                preparedStatement.setString(1, permissionTicketDO.getTicket());
+                preparedStatement.setTimestamp(2, new Timestamp(new Date().getTime()),
+                        permissionTicketDO.getCreatedTime());
+                preparedStatement.setLong(3, permissionTicketDO.getValidityPeriod());
+                preparedStatement.setString(4, permissionTicketDO.getStatus());
+                preparedStatement.setString(5, permissionTicketDO.getTenantId());
+                preparedStatement.execute();
+
+                // Checking if the PT is persisted in the db.
+                long id;
+                try (ResultSet resultSet = preparedStatement.getGeneratedKeys()) {
+                    if (resultSet.next()) {
+                        id = resultSet.getLong(1);
+                    } else {
+                        throw new PermissionDAOException(UMAConstants.ErrorMessages
+                                .ERROR_INTERNAL_SERVER_ERROR_FAILED_TO_PERSIST_PT);
+                    }
+                }
+                for (Resource resource : resourceList) {
+                    try (PreparedStatement resourceIdStatement = connection.prepareStatement(
+                            STORE_PT_RESOURCE_IDS_QUERY)) {
+                        resourceIdStatement.setString(1, resource.getResourceId());
+                        resourceIdStatement.setLong(2, id);
+                        resourceIdStatement.execute();
+
+                        try (ResultSet resultSet = resourceIdStatement.getGeneratedKeys()) {
+                            if (resultSet.next()) {
+                                try (PreparedStatement resourceScopeStatement = connection.prepareStatement
+                                        (STORE_PT_RESOURCE_SCOPES_QUERY)) {
+                                    long resourceId = resultSet.getLong(1);
+                                    for (String scope : resource.getResourceScopes()) {
+                                        resourceScopeStatement.setLong(1, resourceId);
+                                        resourceScopeStatement.setString(2, scope);
+                                        resourceScopeStatement.addBatch();
+                                    }
+                                    resourceScopeStatement.executeBatch();
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            connection.commit();
```

Comment:
We need to close connections, result sets and statements in a finally block. Otherwise leaks may take place which is very harmful in a production system

Reply:
Take this as a general comment and fix all places.

Reply:
Also if the connection is passed from outside to a method don't worry to close them. Basically the principle is close the connection in the place you have created it

Reply:
I have used try with resources thereby is it needed to have a finally block to close connections?